rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java

```diff
@@ -503,33 +503,54 @@ protected void registerPegIn(
-            if (rejectedPeginReasonOptional.isEmpty()) {
-                // This flow should never be reached. There should always be a rejected pegin reason.
-                String message = "Invalid state. No rejected reason was returned from evaluatePegin method";
-                logger.error("[{}}] {}", METHOD_NAME, message);
-                throw new IllegalStateException(message);
+
+        switch (peginProcessAction) {
+            case CAN_BE_REGISTERED -> {
+                logger.debug("[{}] Peg-in is valid, going to register", METHOD_NAME);
+                executePegIn(btcTx, peginInformation, totalAmount);
             }
+            case CAN_BE_REFUNDED -> {
+                // If the peg-in cannot be registered means it should be rejected
+                handleRejectedPegin(btcTx,
+                    peginEvaluationResult);
+                logger.debug("[{}] Refunding to address {} ", METHOD_NAME,
+                    peginInformation.getBtcRefundAddress());
+                generateRejectionRelease(btcTx, peginInformation.getBtcRefundAddress(), rskTxHash,
+                    totalAmount);
+                markTxAsProcessed(btcTx);
+            }
+            case CANNOT_BE_REFUNDED -> {
+                // If the peg-in cannot be registered means it should be rejected
+                RejectedPeginReason rejectedPeginReason = handleRejectedPegin(btcTx,
+                    peginEvaluationResult);
+
+                handleNonRefundablePegin(btcTx, peginInformation.getProtocolVersion(),
+                    rejectedPeginReason);
 
-            RejectedPeginReason rejectedPeginReason = rejectedPeginReasonOptional.get();
-            logger.debug("[{}] Rejected peg-in, reason {}", METHOD_NAME, rejectedPeginReason);
-            eventLogger.logRejectedPegin(btcTx, rejectedPeginReason);
-            if (peginProcessAction == PeginProcessAction.CAN_BE_REFUNDED) {
-                logger.debug("[{}] Refunding to address {} ", METHOD_NAME, peginInformation.getBtcRefundAddress());
-                generateRejectionRelease(btcTx, peginInformation.getBtcRefundAddress(), rskTxHash, totalAmount);
+                if (!activations.isActive(RSKIP459)) {
+                    return;
+                }
+                // Since RSKIP459, rejected peg-ins should be marked as processed
                 markTxAsProcessed(btcTx);
-            } else {
-                logger.debug("[{}] Unprocessable transaction {}.", METHOD_NAME, btcTx.getHash());
-                handleUnprocessableBtcTx(btcTx, peginInformation.getProtocolVersion(), rejectedPeginReason);
             }
         }
     }
 
-    private void handleUnprocessableBtcTx(
+    private RejectedPeginReason handleRejectedPegin(BtcTransaction btcTx,
+        PeginEvaluationResult peginEvaluationResult) {
+        return peginEvaluationResult.getRejectedPeginReason().map(reason -> {
```

Comment:
I don't think it's a good idea to return the rejected peg-in reason here. The method is called `handle` so I would expect it to do something but not to return anything.

The rejected reason is obtained from `PeginEvaluationResult` that is received by param. So it's basically returning something that was passed by parameter. Why can't the caller simply get that info from the object?

Maybe this method should renamed to `logRejectedPegin`? That's what it's doing right?

Reply:
Renamed and rearrange